aws/guardduty.go

```diff
@@ -0,0 +1,173 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/guardduty"
+	common "github.com/cloudposse/posse-cli/common/error"
+	"github.com/sirupsen/logrus"
+)
+
+func getGuardDutyClient(region string, role string) *guardduty.GuardDuty {
+	sess := GetSession()
+	creds := GetCreds(sess, role)
+	guardDutyClient := guardduty.New(sess, &aws.Config{Credentials: creds, Region: &region})
+
+	return guardDutyClient
+}
+
+func enableGuardDutyAdminAccount(client *guardduty.GuardDuty, accountID string) {
+	updateInput := guardduty.EnableOrganizationAdminAccountInput{AdminAccountId: &accountID}
+	client.EnableOrganizationAdminAccount(&updateInput)
+}
+
+// We need to enable GuardDuty in the AWS Organizations Management Account so that it can be added as a member
+// account in AWS GuardDuty's Administrator account. Accounts other than the Management Account don't need to be
+// excplicitly enabled, but the MA does.
+func enableGuardDutyInManagementAccount(client *guardduty.GuardDuty) {
+	_, err := client.CreateDetector(&guardduty.CreateDetectorInput{Enable: aws.Bool(true)})
+	if err != nil {
+		logrus.Error(err)
+	}
+}
+
+func isGuardDutyAdministratorAccountEnabled() bool {
+	return false
+}
+
+func containsGuardDutyAdminAccount(s []*guardduty.AdminAccount, e string) bool {
+	for _, a := range s {
+		if *a.AdminAccountId == e {
+			return true
+		}
+	}
+	return false
+}
+
+func guardDutyAdminAccountAlreadyEnabled(client *guardduty.GuardDuty, accountID string) bool {
+	listInput := guardduty.ListOrganizationAdminAccountsInput{}
+	orgConfig, err := client.ListOrganizationAdminAccounts(&listInput)
+	common.AssertErrorNil(err)
+	if containsGuardDutyAdminAccount(orgConfig.AdminAccounts, accountID) {
+		return true
+	}
+	return false
+}
+
+func logGuardDutyMemberAccounts(memberAccounts []AccountWithEmail) {
+	logrus.Info("  AWS GuardDuty Member accounts:")
+
+	for i := range memberAccounts {
+		logrus.Infof("    %s (%s)", memberAccounts[i].AccountID, memberAccounts[i].Email)
+	}
+}
+
+// addMemberAccount adds an account in the AWS Organization as a member of the GuardDuty Administrator Account
+func addGuardDutyMemberAccounts(client *guardduty.GuardDuty, detectorID string, memberAccounts []AccountWithEmail, administratorAcctID string) {
+	accountDetails := make([]*guardduty.AccountDetail, 0)
+	for i := range memberAccounts {
+		currentAccountID := memberAccounts[i].AccountID
+		currentEmailAddress := memberAccounts[i].Email
+		if currentAccountID != administratorAcctID {
+			accountDetails = append(accountDetails, &guardduty.AccountDetail{AccountId: &currentAccountID, Email: aws.String(currentEmailAddress)})
+		}
+	}
+	input := guardduty.CreateMembersInput{AccountDetails: accountDetails, DetectorId: aws.String(detectorID)}
+	result, err := client.CreateMembers(&input)
+	if err != nil {
+		logrus.Error(err)
```

Comment:
does this stops the executing here?

if `if err != nil` then `result` could be `nil` and the next line would panic.

Reply:
maybe use `else`?

Reply:
83836f1f4f6d9f82d60ca41198c462ccabdd6c2b